plancheck/expect_no_deferred_changes.go

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package plancheck
+
+import (
+	"context"
+	"errors"
+	"fmt"
+)
+
+var _ PlanCheck = expectNoDeferredChanges{}
+
+type expectNoDeferredChanges struct{}
+
+// CheckPlan implements the plan check logic.
+func (e expectNoDeferredChanges) CheckPlan(ctx context.Context, req CheckPlanRequest, resp *CheckPlanResponse) {
+	if len(req.Plan.DeferredChanges) == 0 {
+		return
+	}
+
+	var result []error
+	for _, deferred := range req.Plan.DeferredChanges {
+		resourceAddress := "unknown"
+		if deferred.ResourceChange != nil {
+			resourceAddress = deferred.ResourceChange.Address
+		}
+
+		result = append(result, fmt.Errorf("expected no deferred changes, but resource %q is deferred with reason: %q", resourceAddress, deferred.Reason))
+	}
+
+	resp.Error = errors.Join(result...)
+	if resp.Error != nil {
+		return
+	}
+
+	if !req.Plan.Complete {
+		resp.Error = errors.New("expected plan to be marked as complete, but complete was \"false\", indicating that at least one more plan/apply round is needed to converge.")
+		return
+	}
+}
+
+// ExpectNoDeferredChanges returns a plan check that asserts that there are no deffered changes
+// for any resources in the plan.
+func ExpectNoDeferredChanges() PlanCheck {
```

Comment:
Would it be useful to have a `ExpectNoDeferredChange` plan check assertion for a single resource?

Reply:
Good question! Off the top of my head I can't think of a specific testing flow that would need a negative check like that.

My initial thought is that most tests would follow the general flow of testing a single resource:
- Test step 1: Check a resource is causing a plan to be deferred
- Test step 2: Check there are no more deferred changes (i.e. no more rounds needed)

I'm not sure if there is a workflow where a test may want to end before all deferred changes are satisfied, but maybe more provider developer usage will let us know if that'd be beneficial.